### PR TITLE
Task/ena 4386 seq length check

### PR DESCRIPTION
**Pull request by prabh-t**

### CODE REVIEW

File: src/main/java/uk/ac/ebi/embl/api/validation/check/file/FileValidationCheck.java [hunk 1]
```diff
@@ -211,6 +212,37 @@ protected int getAssemblyLevel(ValidationScope scope)
...
+				throw new ValidationEngineException("COVID-19 genome submissions must have a single chromosome sequence", ReportErrorType.VALIDATION_ERROR);
+			}
+
+			long genomeSize = 0;
+			for (Map.Entry<String, AssemblySequenceInfo> entry : sharedInfo.sequenceInfo.entrySet()) {
+				AssemblySequenceInfo info = entry.getValue();
+
+				if (info.getAssemblyLevel() == 0 || info.getAssemblyLevel() == 1) {
+					if (!sharedInfo.agpEntryNames.contains(entry.getKey())) {
+						genomeSize += info.getSequenceLength();
```
Comment (snathanvj) [inline src/main/java/uk/ac/ebi/embl/api/validation/check/file/FileValidationCheck.java:227]: I think agpEntryName set contains object name of the agp row (first column, i.e name of the scaffold) not the component name(i.e could be contig name or another scaffold name), if so, it always excludes the scaffolds in length calculation.

if the contig is placed in agp then we should not include the contig length towards the genomeLength but we should include the scaffold length(unless it is placed on another scaffold) towards the genomeSize, this is my understanding

File: src/test/java/uk/ac/ebi/embl/api/validation/submission/SubmissionValidationPlanTest.java [hunk 2]
```diff
@@ -479,4 +481,25 @@ public void testGenomeSubmissionwitFastawithValidChromosomeListMultipleTimesInPa
 
 		CompletableFuture.allOf(futs.toArray(new CompletableFuture[futs.size()])).join();
 	}
+
```
Comment (snathanvj) [inline src/test/java/uk/ac/ebi/embl/api/validation/submission/SubmissionValidationPlanTest.java:484]: if possible, can you please add a valid case and assert the expected and actual sequence length are same.

File: src/main/java/uk/ac/ebi/embl/api/validation/check/file/FileValidationCheck.java [hunk 3]
```diff
@@ -211,6 +212,37 @@ protected int getAssemblyLevel(ValidationScope scope)
 		return assemblyLevel;
 	}
 
+	public void covid19SeqLengthCheck() throws ValidationEngineException {
+		if (sharedInfo.assemblyType != null && sharedInfo.assemblyType.equalsIgnoreCase(AssemblyType.COVID_19_OUTBREAK.getValue())) {
+			if (sharedInfo.chromosomeNames.size() > 1) {
+				throw new ValidationEngineException("COVID-19 genome submissions must have a single chromosome sequence", ReportErrorType.VALIDATION_ERROR);
+			}
```
Comment (raskoleinonen) [inline src/main/java/uk/ac/ebi/embl/api/validation/check/file/FileValidationCheck.java:219]: - We can't assume that there would be only a single chromosome. We have ~100 other cases as well, not many but they still exist. Having single chromosome restriction would be possible but we would need to discuss it first with the ENA helpdesk if this is something we could impose.
```
select chromosome_cnt, contig_cnt, scaffold_cnt, count(1)
from gcs_assembly 
where assembly_type = 'COVID-19 outbreak'
group by  chromosome_cnt, contig_cnt, scaffold_cnt;
```

File: src/main/java/uk/ac/ebi/embl/api/validation/check/file/FileValidationCheck.java [hunk 4]
```diff
@@ -211,6 +212,37 @@ protected int getAssemblyLevel(ValidationScope scope)
...
+			System.out.println("Genome size: " + genomeSize);
+			validateCovid19SeqLength(genomeSize);
+		}
+	}
+
+	void validateCovid19SeqLength(long seqLength) throws ValidationEngineException {
+		if(seqLength>MAX_SARS_COV2_SEQUENCE_LENGTH) {
+			throw new ValidationEngineException(String.format("SARS-CoV-2 sequence length should be a max of %dbp.", MAX_SARS_COV2_SEQUENCE_LENGTH), ReportErrorType.VALIDATION_ERROR);
+		}
```
Comment (raskoleinonen) [inline src/main/java/uk/ac/ebi/embl/api/validation/check/file/FileValidationCheck.java:243]: Fine to have this as hardcoded for now. Later we would add an API endpoint into the Webin REST service for Webin-CLI that would be given the assembly type, the taxid (or just the sample_id), and genome length.

File: src/main/java/uk/ac/ebi/embl/api/validation/check/file/FileValidationCheck.java [hunk 5]
```diff
@@ -211,6 +212,37 @@ protected int getAssemblyLevel(ValidationScope scope)
 		return assemblyLevel;
 	}
 
+	public void covid19SeqLengthCheck() throws ValidationEngineException {
+		if (sharedInfo.assemblyType != null && sharedInfo.assemblyType.equalsIgnoreCase(AssemblyType.COVID_19_OUTBREAK.getValue())) {
```
Comment (raskoleinonen) [inline src/main/java/uk/ac/ebi/embl/api/validation/check/file/FileValidationCheck.java:216]: - We should add a check that COVID_19_OUTBREAK genomes have taxid 2697049: If they do not we should give the error: 
```
ERROR: Sample organism must be 'Severe acute respiratory syndrome coronavirus 2' (taxid 2697049) for COVID 19 outbreak genomes.
```
- Let's accept the suggested  'covid19SeqLengthCheck' call as the 1st implementation but leave a comment:
```
// TODO: Genome length check should be generic given the tax id for all species with known expected genome length.
```

File: src/main/java/uk/ac/ebi/embl/api/validation/check/file/FileValidationCheck.java [hunk 6]
```diff
@@ -211,6 +212,37 @@ protected int getAssemblyLevel(ValidationScope scope)
...
+					if (!sharedInfo.agpEntryNames.contains(entry.getKey())) {
+						genomeSize += info.getSequenceLength();
+					}
+				} else if (info.getAssemblyLevel() == 2) {
+					genomeSize += info.getSequenceLength();
+				} else {
+					throw new ValidationEngineException("Unexpected assembly level", ReportErrorType.VALIDATION_ERROR);
+				}
+			}
+			System.out.println("Genome size: " + genomeSize);
```
Comment (raskoleinonen) [inline src/main/java/uk/ac/ebi/embl/api/validation/check/file/FileValidationCheck.java:235]: - We should use validation messages not system out println.

File: src/main/java/uk/ac/ebi/embl/api/validation/check/file/FileValidationCheck.java (hunk 1 above)
Comment (raskoleinonen) [reply src/main/java/uk/ac/ebi/embl/api/validation/check/file/FileValidationCheck.java:227]: - Consider unit tests for genome length computation. Optimally we would have test for: just contigs, just chromosomes, contigs + scaffolds, contigs + scaffolds + chromosomes, contigs + chromosomes.

File: src/test/java/uk/ac/ebi/embl/api/validation/submission/SubmissionValidationPlanTest.java [hunk 7]
```diff
@@ -479,4 +481,25 @@ public void testGenomeSubmissionwitFastawithValidChromosomeListMultipleTimesInPa
...
+	@Test
+	public void testCovid19GenomeSize() throws IOException, ValidationEngineException {
```
Comment (raskoleinonen) [inline src/test/java/uk/ac/ebi/embl/api/validation/submission/SubmissionValidationPlanTest.java:486]: - Totally fine to have two integration tests for covid: one that passed the length validation and another that does not.
- Other than this I would add specific tests for the genome length for the different configuration (e.g. contig only etc.).

File: src/main/java/uk/ac/ebi/embl/api/validation/GenomeUtils.java [hunk 8]
```diff
@@ -0,0 +1,27 @@
+package uk.ac.ebi.embl.api.validation;
+
+import uk.ac.ebi.embl.api.entry.AssemblySequenceInfo;
+
+import java.util.*;
+
+public class GenomeUtils {
```
Comment (raskoleinonen) [inline src/main/java/uk/ac/ebi/embl/api/validation/GenomeUtils.java:7]: - MIght want to have this final with private constuctor.

File: src/main/java/uk/ac/ebi/embl/api/validation/GenomeUtils.java [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+
+    public final static Long COVID_19_OUTBREAK_GENOME_MAX_SIZE = 31000L; //bp
+
+    public static Long calculateGenomeSize(Map<String, AssemblySequenceInfo> sequenceInfo, Set<String> agpPlacedComponents) throws ValidationEngineException {
+        long genomeSize = 0;
+        for (Map.Entry<String, AssemblySequenceInfo> entry : sequenceInfo.entrySet()) {
+            AssemblySequenceInfo info = entry.getValue();
+
+            if (info.getAssemblyLevel() == 0 || info.getAssemblyLevel() == 1) {
+                if (!agpPlacedComponents.contains(entry.getKey())) {
```
Comment (raskoleinonen) [inline src/main/java/uk/ac/ebi/embl/api/validation/GenomeUtils.java:17]: - We might want to uppercase here as well for safety reasons or have unit tests that also check that we can match placed sequences case-insensitively.

File: src/main/java/uk/ac/ebi/embl/api/validation/check/file/FileValidationCheck.java [hunk 10]
```diff
@@ -211,6 +211,18 @@ protected int getAssemblyLevel(ValidationScope scope)
...
 	}
 
+	// TODO: Genome length check should be generic given the tax id for all species with known expected genome length.
+	public void validateCovid19GenomeSize() throws ValidationEngineException {
+		if (sharedInfo.assemblyType != null && sharedInfo.assemblyType.equalsIgnoreCase(AssemblyType.COVID_19_OUTBREAK.getValue())) {
+
+			long genomeSize = GenomeUtils.calculateGenomeSize(sharedInfo.sequenceInfo, sharedInfo.agpPlacedComponents);
+
+			if (genomeSize > GenomeUtils.COVID_19_OUTBREAK_GENOME_MAX_SIZE) {
+				throw new ValidationEngineException(String.format("%s maximum genome size is %d bp.", AssemblyType.COVID_19_OUTBREAK.getValue(), GenomeUtils.COVID_19_OUTBREAK_GENOME_MAX_SIZE), ReportErrorType.VALIDATION_ERROR);
```
Comment (raskoleinonen) [inline src/main/java/uk/ac/ebi/embl/api/validation/check/file/FileValidationCheck.java:221]: The ValidationEngineExceptions are meant to be system errors not user errors. However, what we are generating here is an user and needs to be reported as such back to the user. ValidationEngineExceptions usage is unexpected. What I would expect to see that we add a new error in ValidationResult. Possibly we could use throwValidationCheckException.

File: src/test/java/uk/ac/ebi/embl/api/validation/submission/SubmissionValidationPlanTest.java [hunk 11]
```diff
@@ -479,4 +482,67 @@ public void testGenomeSubmissionwitFastawithValidChromosomeListMultipleTimesInPa
...
+		clearInfoFiles(options.processDir.get());
+		String expectedMsg = String.format("%s maximum genome size is %d bp.", AssemblyType.COVID_19_OUTBREAK.getValue(), GenomeUtils.COVID_19_OUTBREAK_GENOME_MAX_SIZE);
+		SubmissionValidationPlan plan = new SubmissionValidationPlan(options);
+		thrown.expect(ValidationEngineException.class);
+		thrown.expectMessage(expectedMsg);
+		plan.execute();
+	}
+
+	@Test
+	public void testCovid19GenomeSize_ChromoList() throws ValidationEngineException, IOException {
```
Comment (raskoleinonen) [inline src/test/java/uk/ac/ebi/embl/api/validation/submission/SubmissionValidationPlanTest.java:528]: - Fine to have the AGP. Perhaps might be good to test the majority failure case: testCovid19GenomeSizeOneChromsomeFailure

-  testCovid19GenomeSizeContigsScafffoldsSuccess
-  testCovid19GenomeSizeContigsScafffoldsFailure
- testCovid19GenomeSizeOneChromosomeSuccess
- testCovid19GenomeSizeOneChromosomeFailure

File: src/test/java/uk/ac/ebi/embl/api/validation/submission/SubmissionValidationPlanTest.java [hunk 12]
```diff
@@ -479,4 +482,67 @@ public void testGenomeSubmissionwitFastawithValidChromosomeListMultipleTimesInPa
 
 		CompletableFuture.allOf(futs.toArray(new CompletableFuture[futs.size()])).join();
 	}
+
+	@Test
+	public void testCovid19GenomeSize_AGP() throws ValidationEngineException, IOException {
+		String rootPath = "genome/covid19_seq_length_check/";
```
Comment (raskoleinonen) [inline src/test/java/uk/ac/ebi/embl/api/validation/submission/SubmissionValidationPlanTest.java:488]: Maybe:
genome/covid19_seq_length_check/
->
genome/size_check/covid19_contig.fasta
genome/size_check/covid19_scaffolda.agp

genome/size_check/covid19_one_chromosome.fasta
genome/size_check/covid19_one_chromosome_list.txt

**Pull request merged**